modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1949 @@
+     *
+     * To extrace the Vec4d or Vec4f, see toVec()
+     *
+     * To extrace the Rodrigues vector, see toRodrigues()
+     *
+     * If there are two quaternions \f$q_0, q_1\f$ interpolation is needed, you can use nlerp(), slerp() or spline()
+     * ```
+     * cv::nlerp(q0, q1, t)
+     *
+     * cv::slerp(q0, q1, t)
+     *
+     * cv::spline(q0, q0, q1, q1, t)
+     * ```
+     * spline can show the rotation smoothness between multiple quaternions
+     *
+     *
+     * Three ways to get a element in Quaternion
+     * ```
+     * Quatf q(1,2,3,4);
+     * std::cout << q.w << std::endl; // w=1,x=2,y=3,z=4
+     * std::cout << q[0] << std::endl; // q[0]=1, q[1]=2,q[2]=3,q[3]=4
+     * std::cout << q.at(0) << std::endl;
+     * ```
+     */
+    template <typename _Tp> class Quat;
+    template <typename S>
+    std::ostream& operator<<(std::ostream&, const Quat<S>&);
+
+    template <typename _Tp>
+    class Quat
+    {
+    using value_type = typename std::enable_if<
+        std::is_same<float, _Tp>::value||
+        std::is_same<double, _Tp>::value,
+        _Tp
+        >::type;
+    public:
+        Quat() = default;
+
+        /**
+         * @brief From Vec4d or Vec4f
+         */
+        explicit Quat(const cv::Vec<_Tp, 4> &coeff);
+
+        /**
+         * @brief from four numbers
+         */
+        Quat(_Tp w, _Tp x, _Tp y, _Tp z);
+
+        /**
+         * @brief from a angle, axis and qNorm with default 1, axis will be normalized in this constrctor. And
+         * it generates
+         * \f[q = [\cos\theta, u_x\sin\theta,u_y\sin\theta,  u_z\sin\theta].\f]
+         */
+        Quat(const _Tp angle, const cv::Vec<_Tp, 3> &axis);
+
+        /**
+         * @brief from a 3x3 rotate matrix
+         */
+        explicit Quat(const cv::Mat &R);
+
+        /**
+         * @brief from Rodrigues vector
+         */
+        explicit Quat(const Vec<_Tp, 3> &rvec);
+
+        /**
+         * @brief a way to get element
+         * @param index over a range [0, 3].
+         *
+         * Assume a quaternion q,
+         *
+         * q.at(0) is equicalent to q.w,
+         *
+         * q.at(1) is equicalent to q.x,
+         *
+         * q.at(2) is equicalent to q.y,
+         *
+         * q.at(3) is equicalent to q.z,
+         */
+        _Tp at(size_t index) const;
+
+        /**
+         * @brief return the conjugate of this quaternion
+         * \f[q.conjugate() = (w, -x, -y, -z)\f]
+         */
+        Quat<_Tp> conjugate() const;
+
+        /**
+         * @brief return the value of exponential value
+         * \f[\exp(q) = e^w (\cos||\boldsymbol{v}||+ \frac{v}{||\boldsymbol{v}||})\sin||\boldsymbol{v}||\f]
+         * @param q a quaternion
+         *
+         * For example:
+         * ```
+         * Quatd q{1,2,3,4};
+         * cout << exp(q) << endl;
+         * ```
+         */
+        template <typename T>
+        friend Quat<T> exp(const Quat<T> &q);
```

Comment:
It can be implemented without friendship, because there is public method `exp`.

Reply:
I believe this is done for "documentation" purposes.
See: http://pullrequest.opencv.org/buildbot/export/pr/18335/docs/da/d4a/classcv_1_1Quat.html#friends